### PR TITLE
Allow to serve a website via HTTP and HTTPS and don’t redirect HTTP to HTTPS.

### PR DESCRIPTION
**Pull request by ypid**
* This feature makes use of the currently unused False state of the
      `item.redirect_to_ssl` variable and defines it to what I hope is the
      expected behavior.
* This feature is indented for webserver operators who want to offer
      there site via HTTPS for those users who really want HTTPS but does
      not want to get a TLS cert which is signed by a trusted CA.
* This will probably/hopefully get kind of obsolete with
      https://letsencrypt.org/

I can not get the `item.location_list[x].when_schema` to work correctly. The problem is that the `nginx_macro_url_schema` variable defined by the macro `print_shared_nginx_server_block` is not available inside blocks.

There is supposed to be a way by [setting the `scoped` modifier to the block](http://jinja.pocoo.org/docs/dev/templates/#block-nesting-and-scope) but that results in:

    fatal: [jessie.testing] => {'msg': "KeyError: 'undefined variable: 0'", 'failed': True}
    fatal: [jessie.testing] => {'msg': 'One or more items failed.', 'failed': True, 'changed': False, 'results': [{'msg': "KeyError: 'undefined variable: 0'", 'failed': True}]}

Miniexample:

```jinja
{% macro print_shared_nginx_server_block(nginx_macro_url_schema='http') %}
# {{ nginx_macro_url_schema }}
{% block testing scoped %}
# {{ nginx_macro_url_schema }}
{% endblock %}
{% endmacro %}

{{ print_shared_nginx_server_block('https') }}
```

Do you have any idea how to make this work?

### CODE REVIEW

**Comment by ypid** [comment]
All right. Using Nginx conditions is probably the better chose for this instead of `item.location_list[x].when_schema`.

**Comment by drybjed** [comment]
I'm not sure what this patch is supposed to do... Do you want to be able to specify which location entries you want to configure in nginx server block for HTTP or HTTPS independently?

If you want to publish site both on HTTP and HTTPS without explicit redirection to HTTPS, there's already a way to do this:
```
nginx_servers: [ '{{ nginx_server_http }}', '{{ nginx_server_https }}' ]

nginx_server_http:
  name: []
  enabled: True
  ssl: False

nginx_server_https:
  name: []
  enabled: True
  listen: False
```
The first server block will listen on HTTP without adding HTTPS-related configuration. The second server will listen only on HTTPS so that both servers can work basically independently from eachh other. You can still specify shared location using separate dictionary.

One other thing - this patch is kinda big, are all of the changes related to this specific functionality?

**Comment by ypid** [comment]
I saw the need for this feature when working with https://github.com/carlalexander/debops-wordpress
Sure, it could be done entirely in the wordpress role itself but I think it makes sense to allow this via `item.redirect_to_ssl`.

Doing it with two servers directives would be a lot of duplication (https://github.com/carlalexander/debops-wordpress/blob/1f7f5a4a6a6642bc86d2d916e3296069be2eff3b/roles/wordpress/vars/main.yml#L4-L28)

With this feature, it boils down to:

```YAML
wordpress_varnish_server:
  by_role: 'carlalexander.wordpress'
  enabled: True
  default: False
  filename: 'varnish.{{ wordpress_domain }}'
  name: '{{ wordpress_nginx_server_name }}'
  listen: [ '80' ]
  listen_ssl: [ '443' ]
  ssl: '{{ wordpress_ssl }}'
  pki_crt: '{{ wordpress_nginx_crt }}'
  pki_key: '{{ wordpress_nginx_key }}'
  redirect_to_ssl: '{{ wordpress_redirect_to_ssl }}'
  root: False
  index: False
  deny_hidden: False
  favicon: False
  options: |
    client_max_body_size  {{ wordpress_php5_upload_size }};

  location_list:
    - pattern: '/'
      options: |
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        set $X_Forwarded_Proto "http";
        set $X_Forwarded_Port 80;
        if ($https) {
                set $X_Forwarded_Proto "https";
                set $X_Forwarded_Port 443;
        }
        proxy_set_header X-Forwarded-Proto $X_Forwarded_Proto;
        proxy_set_header X-Forwarded-Port $X_Forwarded_Port;
        proxy_set_header Host $host;
        proxy_pass http://varnish;
``` 

Related to: https://github.com/carlalexander/debops-wordpress/issues/1

> One other thing - this patch is kinda big, are all of the changes related to this specific functionality?

Sorry for that. I rebased and put it into #77.

**Comment by drybjed** [comment]
@ypid OK, but now since you dropped the `item.with_schema` code there basically nothing changed apart from moving the location code to separate `{% macro %}` block, right? I'm asking because there are other nginx server types than `default` which use `default` template as a base, and I'm not sure if moving location code to separate macro will affect them.

NB, there's currently a `proxy` template available. I think the proposed `$https`, ie. nginx based solution could be put there, and the wordpress role could be switched to `proxy` without much issues. And I think it even supports the `$https` scenario already. Did you see it?

**Comment by ypid** [comment]
That’s right. I only moved the entire code block inside a macro to avoid duplicating it and I tested it now with the `proxy` template and gone over the other templates. It should not break other things.

Thanks for the hint. The `proxy` template works nicely with updated `$https` condition :+1:

File: templates/etc/nginx/sites-available/default.conf.j2 [hunk 1]
```diff
@@ -111,7 +111,8 @@
 #   - item.redirect_to_ssl: True/False
 #       Optional. If True, redirect connection from HTTP to HTTPS version of
```
Comment (htgoebel) [inline templates/etc/nginx/sites-available/default.conf.j2]: Please document what the default value is. Thanks.

**Pull request merged**